### PR TITLE
hoc2021: Optional SVG animation for overview page

### PR DESCRIPTION
**Pull request by breville**
The overview page at https://code.org/hourofcode/overview features an image with the coloured arrows, added in https://github.com/code-dot-org/code-dot-org/pull/42902.  But now if the `hoc_cse_enable_svg` DCDO flag is set to `true` it's replaced by an SVG which is animated by JavaScript.  

It's disabled by default, at least for now.

The SVG shows the Hour of Code logo immediately, so if the JavaScript doesn't start for some reason, there will still be an image there.

Tested and looks to work correctly in all the regular browsers: Chrome, Safari, Firefox, Edge, and IE11.

Also note that this is deliberately written in vanilla JavaScript rather than going through our build pipeline.  This was done initially in case we used this on one of our homepages, where we want to minimise load time.  The code feels fairly throwaway, since its use will end once we switch to another theme.

### default
Screen Shot 2021-10-13 at 5 28 05 PM

### SVG animation

https://user-images.githubusercontent.com/2205926/137129953-45fa9590-a100-4b8c-a340-963a3e5e7a4e.mov

### CODE REVIEW

File: pegasus/sites.v3/code.org/public/js/cse-bars.js [hunk 1]
```diff
@@ -0,0 +1,490 @@
+var gridSize = 18;
+var strokeWidth = 11;
+var lenTimes = 1;
```
Comment (maddiedierker) [inline pegasus/sites.v3/code.org/public/js/cse-bars.js]: i don't think i understand what this variable means

File: pegasus/sites.v3/code.org/public/js/cse-bars.js [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+var numPanels = 1;
+
+var step = 0;
+
+$(document).ready(function() {
+  var barDefinitions = [];
+
+  barDefinitions[0] = [
```
Comment (maddiedierker) [inline pegasus/sites.v3/code.org/public/js/cse-bars.js:19]: i'm trying to understand the relationship between `barDefinitions` and `numPanels` -- if we were to increment `numPanels`, we'd need to create `barDefinitions[1]`. could we get rid of `numPanels` and instead replace it with `barDefinitions.length`? or is there a reason to keep `numPanels`?

File: pegasus/sites.v3/code.org/public/js/cse-bars.js [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+    // Right side.
+    [
+      {
+        type: "arc",
+        start: [17, 6],
+        end: [18, 5],
+        color: "rgba(144, 140, 203, 0.8)",
+        direction: "right-up"
+      },
```
Comment (maddiedierker) [inline pegasus/sites.v3/code.org/public/js/cse-bars.js:28]: out of curiosity, where did the values in this object come from? did you have to manually map the SVG to this object?

File: pegasus/sites.v3/code.org/public/js/cse-bars.js [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+        color: "rgba(138, 80, 163, 0.8)"
+      },
+      { type: "arrow", color: "rgba(234, 76, 155, 0.8)", direction: "left" }
+    ]
+  ];
+
+  var bars = [];
+  for (var panel = 0; panel < numPanels; panel++) {
+    bars[panel] = [];
+    var panelRef = $(".bars")[panel];
```
Comment (maddiedierker) [inline pegasus/sites.v3/code.org/public/js/cse-bars.js:287]: should we do anything to ensure jQuery is loaded? or is it globally loaded in pegasus somewhere?

File: pegasus/sites.v3/code.org/public/js/cse-bars.js [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+      }
+    }
+    step += 0.02;
+  }, 1000 / 60);
+});
+
+function updateBar(barDefinition, bar, barIndex, step) {
+  for (var segment = 0; segment < barDefinition.length; segment++) {
+    var offset;
+    var phase = 12 - ((step + barIndex + 0.1 * segment) % 12);
```
Comment (maddiedierker) [inline pegasus/sites.v3/code.org/public/js/cse-bars.js:307]: what does 12 mean here?

File: pegasus/sites.v3/code.org/public/js/cse-bars.js [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+    } else if (phase > 12) {
+      offset = 1;
+    } else {
+      offset = 2;
+    }
+
+    var scaledOffset;
+
+    var barDef = barDefinition[segment];
+    if (barDef.type === "arrow") {
```
Comment (maddiedierker) [inline pegasus/sites.v3/code.org/public/js/cse-bars.js]: should we instead check for the presence of head1 and head2 on bar[segment] here? that would decouple this from the implementation in drawBar

**Review by KylieModen — APPROVED** [review]
lgtm - I do think this would be great to include and wonder how we can get the test coverage/confidence to enable this - I saw you already verified on the main browsers, is that enough?

File: pegasus/sites.v3/code.org/public/js/cse-bars.js (hunk 2 above)
Comment (breville) [reply pegasus/sites.v3/code.org/public/js/cse-bars.js:19]: Oh, this was when I was looking at driving two separate SVGs simultaneously to make this work on the other pages.  But you're right that we can just depend on the length of `barDefinitions.length`.

File: pegasus/sites.v3/code.org/public/js/cse-bars.js (hunk 3 above)
Comment (breville) [reply pegasus/sites.v3/code.org/public/js/cse-bars.js:28]: Yeah, I typed this all out by hand by eyeballing the original graphics.  A cooler feature would have been to just state relative direction changes, and even cooler would be to auto-generate paths, but given the rapid approach to this, this is as far as I got.

File: pegasus/sites.v3/code.org/public/js/cse-bars.js (hunk 4 above)
Comment (breville) [reply pegasus/sites.v3/code.org/public/js/cse-bars.js:287]: jQuery is already loaded for the page in question, so I think it's okay to rely on it here.

File: pegasus/sites.v3/code.org/public/js/cse-bars.js (hunk 5 above)
Comment (breville) [reply pegasus/sites.v3/code.org/public/js/cse-bars.js:307]: Ah, just reversing the direction of the numbers.  Once I couldn't rely on negative `stroke-dashoffset` I was looking for a simple way to reverse the flow of the animation without touching most of the logic, and inverting this from 0->~12 to ~12->0 was an easy fix.  Obviously it's all a bit of a hack, but it does work in this limited, temporary scope. 😬

File: pegasus/sites.v3/code.org/public/js/cse-bars.js (hunk 1 above)
Comment (breville) [reply pegasus/sites.v3/code.org/public/js/cse-bars.js]: Same.  It was from prototyping, which wasn't so long ago.  It's gone now :)

**Review by maddiedierker — APPROVED** [review]
🎉 SVG animations are so fun!

**Comment by breville** [comment]
@madelynkasula Thanks very much for the detailed review.  If this code were any longer-lived or more widely-used I'd give it a complete rewrite, but fingers crossed it works for this fun graphic for now. :)

**Pull request merged**